### PR TITLE
add perfdash in CDK

### PR DESCRIPTION
**Pull request by mengqiy**
Added perfdash components (excluding configmap) in CDK.
Added a section in the README for perfdash.

### CODE REVIEW

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+kind: ConfigMap
+metadata:
+  name: perfdash-config
+  namespace: perfdash
+  labels:
+    app.kubernetes.io/name: perfdash
+    app.kubernetes.io/component: config
+    app.kubernetes.io/part-of: perfdash
+data:
+  jobs.yaml: |
```
Comment (hakuna-matatah) [inline infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:16]: - This data about jobs are more specific to internal things that we run. It doesn't make sense to keep this in upstream repo. Could we push this in our internal repo.
wdyt ?

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        app.kubernetes.io/part-of: perfdash
+    spec:
+      containers:
+        - name: perfdash
+          image: gcr.io/k8s-staging-perf-tests/perfdash:2.44
+          command:
+            - /perfdash
+            - --www=true
+            - --address=0.0.0.0:8080
+            - --builds=20
```
Comment (hakuna-matatah) [inline infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:75]: IIUC does this mean we only display 20 builds on dashboard ?

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            - --force-builds
+            - --mode=s3
+            - --aws-region=us-west-2
+            - --configPath=/etc/config/jobs.yaml
+            - --logsBucket=kit-eks-scalability
```
Comment (hakuna-matatah) [inline infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:80]: Either we need to make sure we explicitly document this detail because other customers outside of us won't have access to this bucket and this deployment pods might fail or throw errors in the logs (OR) move this to internal repo ? Or patch this specific detail through flux  by taking in as param through cdk command line args ?

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            - --logsPath=
```
Comment (hakuna-matatah) [inline infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:81]: any reason for having this here with empty value ?

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 1 above)
Comment (mengqiy) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:16]: That makes sense.

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 2 above)
Comment (mengqiy) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:75]: What's your suggestion? 50 or 100 or some other value?

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 4 above)
Comment (mengqiy) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:81]: It appears upstream has another layer called `logs` in the hierarchy. We don't have that and it's not necessary for our use case. So we have it as an empty value.

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 3 above)
Comment (mengqiy) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:80]: I don't quite understand this comment.
Who is other customers? Other EKS teams or EKS customers?
IIUC this is bucket is only accessible to our team. Others can't access it no matter where we keep this yaml file.
Can you please elaborate your concern? We can discuss offline if you want.

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 4 above)
Comment (hakuna-matatah) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:81]: cool, i see it here - https://github.com/kubernetes/perf-tests/blob/master/perfdash/perfdash.go#L52

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml (hunk 3 above)
Comment (mengqiy) [reply infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:43]: Updated according to our offline discussion.

**Comment by mengqiy** [comment]
PTAL

**Comment by mengqiy** [comment]
@hakuna-matatah Please note that I have created an internal CR for the ConfigMap. We need to merge and apply that before we merge this PR.

File: infrastructure/README.md [hunk 5]
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
...
+Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, the metrics-server and [perf dash].
+
+#### Perf Dash
+
+Perf dash is optional. If you want to disable it, you can simply remove or comment out the content in `./infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml`.
+
+To bring it up, there are a few places you need to ensure they are wired correctly.
+
+- The `logsBucket` logs bucket should be your bucket holding the logs. Ideally, it should be in the same AWS account as your KIT Infrastructure. Otherwise, you may need to configure the auth yourself.
+- A ConfigMap containing the jobs config for perf dash must be created so that it can be mounted as volumn in the Deployment. An example configmap can be:
```
Comment (hakuna-matatah) [inline infrastructure/README.md:30]: Nit: volum -> volume

File: infrastructure/README.md [hunk 6]
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
...
-Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, and the metrics-server. 
+Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, the metrics-server and [perf dash].
+
+#### Perf Dash
+
+Perf dash is optional. If you want to disable it, you can simply remove or comment out the content in `./infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml`.
+
+To bring it up, there are a few places you need to ensure they are wired correctly.
+
+- The `logsBucket` logs bucket should be your bucket holding the logs. Ideally, it should be in the same AWS account as your KIT Infrastructure. Otherwise, you may need to configure the auth yourself.
```
Comment (hakuna-matatah) [inline infrastructure/README.md:29]: Thanks for adding these details.

File: infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        app.kubernetes.io/part-of: perfdash
+    spec:
+      containers:
+        - name: perfdash
+          image: gcr.io/k8s-staging-perf-tests/perfdash:2.44
+          command:
+            - /perfdash
+            - --www=true
+            - --address=0.0.0.0:8080
+            - --builds=100
```
Comment (hakuna-matatah) [inline infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:38]: Nice!

**Comment by hakuna-matatah** [comment]
> @hakuna-matatah Please note that I have created an internal CR for the ConfigMap. We need to merge and apply that before we merge this PR.
> 
> 

Yup, thanks for moving that to internal.

File: infrastructure/README.md (hunk 5 above)
Comment (mengqiy) [reply infrastructure/README.md]: Done

**Comment by mengqiy** [comment]
@hakuna-matatah  PTAL

**Review by hakuna-matatah — DISMISSED** [review]
lgtm

**Comment by mengqiy** [comment]
This PR should hold merging until the ConfigMap has been created out-of-the-band.

File: infrastructure/README.md [hunk 8]
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
 - Flux v2
 - Kubernetes Iteration Toolkit (KIT) Operator (optional)
 
-Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, and the metrics-server. 
+Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, the metrics-server and [perf dash].
```
Comment (prateekgogia) [inline infrastructure/README.md:21]: Curious why do we have perf dash text formatting like this - `[perf dash]`?

File: infrastructure/README.md [hunk 9]
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
...
+
+#### Perf Dash
+
+Perf dash is optional. If you want to disable it, you can simply remove or comment out the content in `./infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml`.
```
Comment (prateekgogia) [inline infrastructure/README.md:25]: Doesn't provide info if its enabled by default? 
> If you want to disable it, you can simply remove or comment out the content in

I don't think this will work because it will be synced from Flux and we don't want everyone to be changing KIT repo for disabling perfdash. 

Also enabling by default would mean existing clusters in Karpenter will get this installed which might be a surprise for them. I would minimize such changes in this repo or at least get a consent from Karpenter team.
cc @bwagner5

File: infrastructure/README.md [hunk 10]
```diff
@@ -69,4 +115,6 @@ cdk deploy KITInfrastructure --no-rollback \
 | TestServiceAccount  | Service Account for application tests to run with                                          |                                                         |   |   |
 | KITAddon            | KIT CRD addon that gets installed on KIT Infrastructure by default                         | true                                                    |   |   |                               
 | KarpenterAddon      | Karpenter CRD addon that gets installed on KIT Infrastructure by default                   | true                                                    |   |   | 
-| AWSEBSCSIDriverAddon| AWSEBSCSIDriver CRD addon that gets installed on KIT Infrastructure by default             | true                                                    |   |   |
\ No newline at end of file
+| AWSEBSCSIDriverAddon| AWSEBSCSIDriver CRD addon that gets installed on KIT Infrastructure by default             | true                                                    |   |   |
+
+[perf dash]: https://github.com/kubernetes/perf-tests/tree/master/perfdash
```
Comment (prateekgogia) [inline infrastructure/README.md:120]: Why different formatting?

**Review by prateekgogia — DISMISSED** [review]

File: infrastructure/README.md (hunk 8 above)
Comment (mengqiy) [reply infrastructure/README.md:21]: It makes the raw markdown easier to read when the long URLs is not embedded in the text.
The link of `perf dash` is defined at the bottom of this file.

File: infrastructure/README.md (hunk 9 above)
Comment (mengqiy) [reply infrastructure/README.md:25]: Thanks for bringing this up! It seems karpenter do use KIT for testing: https://github.com/aws/karpenter/tree/main/test/infrastructure#setup.

>I don't think this will work because it will be synced from Flux and we don't want everyone to be changing KIT repo for disabling perfdash.

I thought a user cloud have their own forks where they can make some changes. I can rephrase it to make it more clear. If you don't think that's a valid assumption, I can change it.

>Also enabling by default would mean existing clusters in Karpenter will get this installed which might be a surprise for them.

I guess it won't be a surprise for Karpenter team, since it doesn't break their cluster. But I guess this addon is not useful for them, since they don't CL2 tests at all. We may not want to install it default.
An alternative is to add it as an addon under `./infrastructure/lib/addons/` using typescript. A flag can be used to gate it and make it disabled by default.
@prateekgogia @hakuna-matatah WDYT?

File: infrastructure/README.md (hunk 10 above)
Comment (mengqiy) [reply infrastructure/README.md:120]: Answered in https://github.com/awslabs/kubernetes-iteration-toolkit/pull/306#discussion_r1009719638

File: infrastructure/README.md (hunk 9 above)
Comment (hakuna-matatah) [reply infrastructure/README.md:25]: Current state of things - We install certain addons today through typescript cdk thing if that addon requires aws resources to function otherwise we install them through flux. Customers have an option of disabling certain addons through cdk if they are being installed but not through flux.
We need to take a stand on whether we want to move all the addons to be installed through type-script and give customers an option of disabling unwanted addons (OR) have a separate directory or git repo to be pointed out for all optional addons so customers have flexibility to choose at the time of setup. For the latter one (i.e; flux route), all we have to do is basically send an array list here for repoPaths and repos, so customers can pass it at the time of setup  - https://github.com/awslabs/kubernetes-iteration-toolkit/blob/main/infrastructure/lib/kit-infrastructure.ts#L167-L177

Comment (mengqiy) [reply infrastructure/README.md:25]: > have a separate directory or git repo to be pointed out for all optional addons so customers have flexibility to choose at the time of setup.

I guess this approach makes more sense.

**Comment by mengqiy** [comment]
Rebased on top of the latest HEAD. The last commit is the change since last review.
The changes is not fully tested due to https://github.com/awslabs/kubernetes-iteration-toolkit/issues/313. I will need some help to figure out what's going on with #313.

File: infrastructure/lib/addons/fluxv2.ts [hunk 11]
```diff
@@ -10,10 +10,12 @@ export interface FluxV2Props extends StackProps {
   repoUrl: string;
   repoBranch: string;
   repoPath: string;
+  repoAddonPaths?: string[];
   testRepoName: string;
   testRepoUrl?: string;
   testRepoBranch?: string;
   testRepoPath?: string;
+  testRepoAddonPaths?: string[];
```
Comment (hakuna-matatah) [inline infrastructure/lib/addons/fluxv2.ts:18]: why would you  need two arrays like one [here](https://github.com/awslabs/kubernetes-iteration-toolkit/pull/306/files#diff-bbf436cf80554d1c16a136a08cbc43d87d488d7ab9d51b6ca1aa66a74de555b9R13) and on this line ? Can't we just have one.

File: infrastructure/lib/addons/fluxv2.ts [hunk 12]
```diff
@@ -127,6 +129,30 @@ export class FluxV2 extends Construct {
   });
   kustomizationManifest.node.addDependency(chart);
 
+  if (props.repoAddonPaths) {
+    for (var repoAddonPath of props.repoAddonPaths) {
+      const addonKustomizationManifest = props.cluster.addManifest('AddonKustomizationSelf', {
```
Comment (hakuna-matatah) [inline infrastructure/lib/addons/fluxv2.ts:134]: could we wrap this in `function` as we are using it multiple times ?

File: infrastructure/lib/addons/fluxv2.ts (hunk 11 above)
Comment (mengqiy) [reply infrastructure/lib/addons/fluxv2.ts:18]: From my current understanding, we support `Repo` and `testRepo` (this is not a great name, but we are using it now), user may want to install the addons from base `Repo` and (or) `testRepo`. We should provide both options IMO.

File: infrastructure/lib/addons/fluxv2.ts (hunk 12 above)
Comment (mengqiy) [reply infrastructure/lib/addons/fluxv2.ts:134]: That's possible.
https://github.com/awslabs/kubernetes-iteration-toolkit/blob/da36db42fe000ace00c9396fb3abd0af226da3b0/infrastructure/lib/addons/fluxv2.ts#L110-L128
and https://github.com/awslabs/kubernetes-iteration-toolkit/blob/da36db42fe000ace00c9396fb3abd0af226da3b0/infrastructure/lib/addons/fluxv2.ts#L150-L168 are doing similar things for `Repo` and `testRepo`. We should clean them up as well eventually.

File: infrastructure/lib/addons/fluxv2.ts (hunk 11 above)
Comment (hakuna-matatah) [reply infrastructure/lib/addons/fluxv2.ts:18]: The idea there was to have `repo` for k8s addons that needs to be installed on Management cluster through Flux and `testRepo` is a place holder for tekton tasks tests etc that we wanted to separate out at that time from the Infrastructure code repository itself. 


Given with this PR, now we can take in array of repos, this logical separation of `repos` and `testRepos` may not be needed at this point. Because customers can bring in any number of arbitrary set of `testrepos` which contain tests to sync on cluster and/or also arbitrary set of k8s addons can be installed  `repos` using the same array list . Doesn't add any benefit now to have separate vars for this. Makes sense ?

**Comment by mengqiy** [comment]
@hakuna-matatah PTAL

**Review by hakuna-matatah — APPROVED** [review]
lgtm
